lib/usagereporter/usagereporter.go

```diff
@@ -230,19 +240,50 @@ func (r *UsageReporter[T]) enqueueBatch() {
 	}
 }
 
+// GracefulStop stops receiving new events and schedules the
+// final batch for submission. It blocks until the final batch
+// has been sent, or until the provided context is canceled.
+func (r *UsageReporter[T]) GracefulStop(ctx context.Context) error {
+	wait := make(chan struct{})
+	go func() {
+		r.wg.Wait()
+		close(wait)
+	}()
+
+	close(r.gracefulStopScheduledC)
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-wait:
+		return nil
+	}
+}
+
 // Run begins processing incoming usage events. It should be run in a goroutine.
 func (r *UsageReporter[T]) Run(ctx context.Context) {
+	r.wg.Add(1)
+	defer r.wg.Done()
+
+	submitCtx, cancelSubmit := context.WithCancel(ctx)
+
 	timer := r.clock.NewTimer(r.maxBatchAge)
 
 	// Also start the submission goroutine.
-	go r.runSubmit(ctx)
+	go r.runSubmit(submitCtx)
 
 	r.Debug("usage reporter is ready")
 
 	for {
 		select {
 		case <-ctx.Done():
 			return
+		case <-r.gracefulStopScheduledC:
+			r.mtx.Lock()
+			r.wasGracefulStopScheduled = true
+			r.mtx.Unlock()
+			r.enqueueBatch()
+			cancelSubmit() // it will stop runSubmit loop after a pending submit finishes
```

Comment:
In addition to having no guarantee that the leftover events fit in a single batch (which can be an explicit choice, "we will only submit up to one batch before shutting down"), you also have no guarantee that the `runSubmit` goroutine will wake up after the batch was submitted but before the context was canceled.

Reply:
> you also have no guarantee that the runSubmit goroutine will wake up after the batch was submitted but before the context was canceled.

Is this because `submissionQueue` is buffered?